File: scripts/check-coverage.ts
import { isAbsolute, relative, resolve } from "path";

const THRESHOLD = 95;

type CoverageConfig = {
  roots: string[];
  extensions: string[];
  reportBase?: string;
  exclude?: string[];
};

export type CoverageResult = {
  threshold: number;
  covered: number;
  total: number;
  percent: number;
  missing: string[];
  passed: boolean;
};

function globMatches(value: string, pattern: string): boolean {
  const escaped = pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&");
  return new RegExp(
    `^${escaped.replaceAll("*", ".*").replaceAll("?", ".")}$`,
  ).test(value);
}

function isInScope(path: string, config: CoverageConfig): boolean {
  const rooted = config.roots.some((prefix) => {
    const normalized = prefix.replace(/\/+$/, "");
    return (
      normalized === "." ||
      path === normalized ||
      path.startsWith(`${normalized}/`)
    );
  });
  return (
    rooted &&
    !(config.exclude ?? []).some((pattern) => globMatches(path, pattern))
  );
}

function trackedFiles(root: string): string[] {
  const result = Bun.spawnSync(["git", "ls-files", "-z"], { cwd: root });
  if (result.exitCode !== 0) {
    throw new Error("Unable to list tracked source files");
  }
  return new TextDecoder()
    .decode(result.stdout)
    .split("\0")
    .filter(Boolean)
    .map((path) => path.replaceAll("\\", "/"));
}

function reportSourcePath(
  source: string,
  root: string,
  reportBase: string,
): string {
  const absolute = isAbsolute(source)
    ? resolve(source)
    : resolve(root, reportBase, source);
  const relativePath = relative(resolve(root), absolute).replaceAll("\\", "/");
  return relativePath === ".." || relativePath.startsWith("../")
    ? ""
    : relativePath;
}

export async function checkCoverage(
  config: CoverageConfig,
  reports: string[],
  root: string,
): Promise<CoverageResult> {
  const sources = new Set(
    trackedFiles(root).filter(
      (path) =>
        config.extensions.some((extension) => path.endsWith(extension)) &&
        isInScope(path, config),
    ),
  );
  if (sources.size === 0) {
    throw new Error("No source files matched the coverage scope");
  }

  const lines = new Map<string, boolean>();
  const seen = new Set<string>();
  for (const report of reports) {
    let current: string | undefined;
    let reportLines = 0;
    const contents = await Bun.file(report).text();
    for (const entry of contents.split(/\r?\n/)) {
      if (entry.startsWith("SF:")) {
        const name = reportSourcePath(
          entry.slice(3),
          root,
          config.reportBase ?? ".",
        );
        current = sources.has(name) ? name : undefined;
        if (current) seen.add(current);
      } else if (entry.startsWith("DA:") && current) {
        const [lineNumber, hitCount] = entry.slice(3).split(",");
        if (
          !/^-?\d+$/.test(lineNumber ?? "") ||
          !/^-?\d+$/.test(hitCount ?? "")
        ) {
          throw new Error("Invalid LCOV line data");
        }
        const number = Number(lineNumber);
        const hits = Number(hitCount);
        if (number <= 0 || hits < 0) {
          throw new Error("Invalid LCOV line data");
        }
        reportLines += 1;
        const key = `${current}\0${number}`;
        lines.set(key, (lines.get(key) ?? false) || hits > 0);
      } else if (entry === "end_of_record") {
        current = undefined;
      }
    }
    if (reportLines === 0) {
      throw new Error(`No scoped line data in ${report}`);
    }
  }

  if (lines.size === 0) {
    throw new Error("No instrumented source lines in coverage reports");
  }
  const missing = [...sources].filter((path) => !seen.has(path)).sort();
  const covered = [...lines.values()].filter(Boolean).length;
  const total = lines.size;
  return {
    threshold: THRESHOLD,
    covered,
    total,
    percent: Math.round((covered * 10000) / total) / 100,
    missing,
    passed: missing.length === 0 && covered * 100 >= THRESHOLD * total,
  };
}

async function main(args: string[]): Promise<number> {
  const [configPath, ...reports] = args;
  if (!configPath || reports.length === 0) {
    console.error(
      "Usage: bun run scripts/check-coverage.ts <config> <lcov-report> [...]",
    );
    return 2;
  }
  const root = process.cwd();
  try {
    const config = (await Bun.file(
      resolve(root, configPath),
    ).json()) as CoverageConfig;
    const result = await checkCoverage(config, reports, root);
    console.log(JSON.stringify(result, null, 2));
    if (result.missing.length > 0) {
      console.error(
        "Coverage is incomplete: scoped source files are absent from LCOV.",
      );
    }
    return result.passed ? 0 : 1;
  } catch (error) {
    console.error(
      `Coverage gate failed: ${error instanceof Error ? error.message : String(error)}`,
    );
    return 1;
  }
}

if (import.meta.main) {
  process.exitCode = await main(process.argv.slice(2));
}

File: scripts/check-coverage.py
"""Fail closed on missing source reports or weighted line coverage below 95%."""
import argparse
import fnmatch
import json
from pathlib import Path
import subprocess
import sys

THRESHOLD = 95


def check(config, reports, root):
    def included(path):
        return any(path == prefix or path.startswith(prefix.rstrip('/') + '/') or prefix == '.'
                   for prefix in config['roots']) and not any(
                       fnmatch.fnmatch(path, pattern) for pattern in config.get('exclude', []))

    sources = {path for path in subprocess.check_output(
        ['git', 'ls-files', '-z'], cwd=root).decode().split('\0')
        if path and Path(path).suffix in config['extensions'] and included(path)}
    if not sources:
        raise ValueError('No source files matched the coverage scope')
    lines = {}
    seen = set()
    for report in reports:
        current = None
        report_lines = 0
        for entry in Path(report).read_text().splitlines():
            if entry.startswith('SF:'):
                path = Path(entry[3:])
                if not path.is_absolute():
                    path = root / config.get('reportBase', '.') / path
                try:
                    name = path.resolve().relative_to(root.resolve()).as_posix()
                except ValueError:
                    name = ''
                current = name if name in sources else None
                if current:
                    seen.add(current)
            elif entry.startswith('DA:') and current:
                fields = entry[3:].split(',')
                number, hits = int(fields[0]), int(fields[1])
                if number <= 0 or hits < 0:
                    raise ValueError('Invalid LCOV line data')
                report_lines += 1
                key = (current, number)
                lines[key] = lines.get(key, False) or hits > 0
            elif entry == 'end_of_record':
                current = None
        if not report_lines:
            raise ValueError(f'No scoped line data in {report}')
    if not lines:
        raise ValueError('No instrumented source lines in coverage reports')
    missing = sorted(sources - seen)
    covered = sum(lines.values())
    total = len(lines)
    passed = not missing and covered * 100 >= THRESHOLD * total
    result = {'threshold': THRESHOLD, 'covered': covered, 'total': total,
              'percent': round(covered * 100 / total, 2), 'missing': missing, 'passed': passed}
    print(json.dumps(result, indent=2))
    if missing:
        print('Coverage is incomplete: scoped source files are absent from LCOV.', file=sys.stderr)
    return passed


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('config')
    parser.add_argument('reports', nargs='+')
    args = parser.parse_args()
    root = Path.cwd()
    try:
        return 0 if check(json.loads(Path(args.config).read_text()), args.reports, root) else 1
    except (OSError, ValueError, KeyError, subprocess.CalledProcessError) as error:
        print(f'Coverage gate failed: {error}', file=sys.stderr)
        return 1


if __name__ == '__main__':
    sys.exit(main())
